glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,543 @@
+                )
+                .expect("Unable to create StallDetectorGuard, giving up")
+            })
+    }
+
+    pub(crate) fn arm(&self, threshold: Duration) -> nix::Result<()> {
+        self.timer.set(
+            sys::timerfd::Expiration::OneShot(sys::time::TimeSpec::from(threshold)),
+            sys::timerfd::TimerSetTimeFlags::empty(),
+        )
+    }
+
+    pub(crate) fn disarm(&self) -> nix::Result<()> {
+        self.timer.unset()
+    }
+}
+
+impl Drop for StallDetector {
+    fn drop(&mut self) {
+        signal_hook::low_level::unregister(self.signal_id);
+        let timer_handler = self.timer_handler.take().unwrap();
+        self.terminated.store(true, Ordering::Relaxed);
+
+        self.timer
+            .set(
+                sys::timerfd::Expiration::Interval(sys::time::TimeSpec::from(
+                    Duration::from_millis(1),
+                )),
+                sys::timerfd::TimerSetTimeFlags::empty(),
+            )
+            .expect("failed wake the timer for termination");
+
+        let _ = timer_handler.join();
+    }
+}
+
+pub(crate) struct StallDetectorGuard<'detector> {
+    detector: &'detector StallDetector,
+    queue_handle: TaskQueueHandle,
+    queue_name: String,
+    start: Instant,
+    threshold: Duration,
+}
+
+impl<'detector> StallDetectorGuard<'detector> {
+    fn new(
+        detector: &'detector StallDetector,
+        queue_handle: TaskQueueHandle,
+        queue_name: String,
+        start: Instant,
+        threshold: Duration,
+    ) -> nix::Result<Self> {
+        detector
+            .arm(threshold)
+            .expect("Unable to arm stall detector, giving up");
+        Ok(Self {
+            detector,
+            queue_handle,
+            queue_name,
+            start,
+            threshold,
+        })
+    }
+}
+
+impl<'detector> Drop for StallDetectorGuard<'detector> {
+    fn drop(&mut self) {
+        let _ = self.detector.disarm();
+
+        let mut frames = vec![];
+        while let Ok(frame) = self.detector.rx.try_recv() {
+            frames.push(frame);
+        }
+        let mut strace = backtrace::Backtrace::from(frames);
+
+        if strace.frames().is_empty() {
+            return;
+        }
+
+        let elapsed = self.start.elapsed();
+        strace.resolve();
+        self.detector.stall_handler.stall(StallDetection {
+            executor: self.detector.id,
+            queue_name: &self.queue_name,
+            queue_handle: self.queue_handle,
+            trace: strace,
+            budget: self.threshold,
+            overage: elapsed.saturating_sub(self.threshold),
+        });
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        executor::{
+            stall::{StallDetection, StallDetectionHandler},
+            TaskQueueHandle,
+        },
+        timer::sleep,
+        LocalExecutorBuilder,
+    };
+    use std::{
+        sync::{Arc, RwLock},
+        thread,
+        time::Duration,
+    };
+
+    #[allow(dead_code)]
```

Comment:
I don't think this is necessary?

Reply:
Not all of the fields are used by tests currently. I can remove the fields if it's a problem?

Reply:
If you plan to add tests that use this then let's wait on those; otherwise, let's remove it :)